### PR TITLE
`build/pkgs/flint` 3.1.3

### LINKED ISSUE #37838: restricting Flint version should be conditional on installing Singular
### Steps To Reproduce

the current 10.4.beta3 does not allow external Singular, because the external flint is too new. #37495 is placing an unconditional upper bound on flint's version, for no good reason.

Users complain, rightly:  https://groups.google.com/g/sage-devel/c/7RKSyswB2VA/m/T-xbNnW5BAAJ
Developers waste time, too.

### Expected Behavior

External Singular should be allowed even even Flint's version is too new to build Sage's Singular, these are different things.

### Actual Behavior

external Singular rejected due to Flint too new.

### Additional Information

_No response_

### Environment

```markdown
- **OS**: Ubuntu 22
- **Sage Version**: 10.4.beta3
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by mkoeppe**
For this upgrade we tighten the check for GMP so that 6.2.1 is accepted but 6.2.0 is rejected. For example `ubuntu-focal-standard` brings 6.2.0, we accept it, and then FLINT complains:
https://github.com/mkoeppe/flint2/actions/runs/7718792281/job/21040793706#step:11:5395

We also change the upper bound on system `flint` accepted by `configure`. Closes #37838.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

- Depends on https://github.com/sagemath/sage/pull/37484
- Depends on #37495 (merged and reverted here)
- Depends on #37570 (merged here)
- Depends on #37351 (merged here)
- Depends on #37726 (merged here to resolve merge conflict)

### CODE REVIEW

**Comment by mkoeppe** [comment]
The tests haven't run yet, but apparently we need a patch for FLINT 3.1.0 support? https://groups.google.com/g/sage-support/c/72w_C1O6lPs/m/QC6M0CtVEAAJ @antonio-rojas

**Comment by antonio-rojas** [comment]
> The tests haven't run yet, but apparently we need a patch for FLINT 3.1.0 support? https://groups.google.com/g/sage-support/c/72w_C1O6lPs/m/QC6M0CtVEAAJ @antonio-rojas

Yes, I will submit a PR in a few hours

**Comment by antonio-rojas** [comment]
Note that singular also needs patching (my PR will only handle sagelib)

https://gitlab.archlinux.org/archlinux/packaging/packages/singular/-/blob/main/flint-3.1.patch?ref_type=heads

**Comment by antonio-rojas** [comment]
> > The tests haven't run yet, but apparently we need a patch for FLINT 3.1.0 support? https://groups.google.com/g/sage-support/c/72w_C1O6lPs/m/QC6M0CtVEAAJ @antonio-rojas
> 
> Yes, I will submit a PR in a few hours

https://github.com/sagemath/sage/pull/37484

**Comment by mkoeppe** [comment]
The incremental build (Build&Test) shows:
```
ImportError: libflint.so.18: cannot open shared object file: No such file or directory
```
(https://github.com/sagemath/sage/actions/runs/8626409427/job/23644543324?pr=37203#step:14:54)

**Comment by mkoeppe** [comment]
Full tests ran at https://github.com/mkoeppe/sage/actions/runs/8761823524

File: build/pkgs/gmp/spkg-configure.m4 [hunk 1]
```diff
@@ -2,6 +2,26 @@ SAGE_SPKG_CONFIGURE([gmp], [
            sage_spkg_install_gmp=no
             AC_CHECK_HEADER(gmp.h, [], [sage_spkg_install_gmp=yes])
             AC_CHECK_HEADER(gmpxx.h, [], [sage_spkg_install_gmp=yes])
+            # Check that GMP >= 6.2.1; this test is copied from FLINT.
```
Comment (kwankyu) [inline build/pkgs/gmp/spkg-configure.m4:5]: Is this alignment correct? See line 2 and line 3.

File: build/pkgs/flint/checksums.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 tarball=flint-VERSION.tar.gz
-sha1=65be9297c06edd7e24f20874b7bd6130cee56723
-md5=5189f67b0ec12e4a54d6782851642b81
-cksum=172350473
+sha1=35af6e2de8765a841952438f3c915d0505b983b6
+md5=f449b4902c51b03e902b2b8bc6ef1307
+cksum=1303232242
```
Comment (kwankyu) [inline build/pkgs/flint/checksums.ini]: `sha256`?

Comment (mkoeppe) [reply build/pkgs/flint/checksums.ini]: done in e2d2df4

File: build/pkgs/gmp/spkg-configure.m4 (hunk 1 above)
Comment (mkoeppe) [reply build/pkgs/gmp/spkg-configure.m4:5]: improved in 090a5af531f

**Comment by kwankyu** [comment]
There is also a merge conflict with https://github.com/sagemath/sage/pull/38025.

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thanks!

**Pull request merged**